distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java

```diff
@@ -590,62 +620,43 @@ void jarHellCheck(PluginInfo info, Path candidate, Path pluginsDir, Path modules
-            Files.walkFileTree(destination, new SimpleFileVisitor<Path>() {
-                @Override
-                public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) throws IOException {
-                    if ("bin".equals(file.getParent().getFileName().toString())) {
-                        setFileAttributes(file, BIN_FILES_PERMS);
-                    } else {
-                        setFileAttributes(file, PLUGIN_FILES_PERMS);
+            if (MetaPluginInfo.isMetaPlugin(tmpRoot)) {
+                final MetaPluginInfo metaInfo = MetaPluginInfo.readFromProperties(tmpRoot);
+                verifyPluginName(env.pluginsFile(), metaInfo.getName(), tmpRoot);
+                final Path metaPath = env.pluginsFile().resolve(metaInfo.getName());
+                terminal.println(VERBOSE, metaInfo.toString());
+
+                // creates the meta plugin directory where all plugins are copied
+                Files.createDirectory(metaPath);
+                deleteOnFailure.add(metaPath);
+                setFileAttributes(metaPath, PLUGIN_DIR_PERMS);
+
+                // copy the meta plugin properties file in the meta plugin directory
+                Path metaInfoDest = metaPath.resolve(ES_META_PLUGIN_PROPERTIES);
+                Files.copy(tmpRoot.resolve(ES_META_PLUGIN_PROPERTIES), metaInfoDest);
+                setFileAttributes(metaInfoDest, PLUGIN_FILES_PERMS);
+
+                // install all plugins
+                try (DirectoryStream<Path> subPaths = Files.newDirectoryStream(tmpRoot)) {
+                    for (Path subPlugin : subPaths) {
+                        if (MetaPluginInfo.isPropertiesFile(subPlugin)) {
+                            continue;
+                        }
+                        installPlugin(terminal, isBatch, subPlugin, env.pluginsFile().resolve(metaInfo.getName()),
+                            env.binFile().resolve(metaInfo.getName()), env.configFile().resolve(metaInfo.getName()), env, deleteOnFailure);
                     }
-                    return FileVisitResult.CONTINUE;
-                }
-
-                @Override
-                public FileVisitResult postVisitDirectory(final Path dir, final IOException exc) throws IOException {
-                    setFileAttributes(dir, PLUGIN_DIR_PERMS);
-                    return FileVisitResult.CONTINUE;
-                }
-            });
-
-            if (info.requiresKeystore()) {
-                KeyStoreWrapper keystore = KeyStoreWrapper.load(env.configFile());
-                if (keystore == null) {
-                    terminal.println("Elasticsearch keystore is required by plugin [" + info.getName() + "], creating...");
-                    keystore = KeyStoreWrapper.create(new char[0]);
-                    keystore.save(env.configFile());
                 }
+                // clean up installation
+                IOUtils.rm(tmpRoot);
```

Comment:
SHouldn't this cleanup and printing be down outside of the if/else since it should apply to both types of plugins?

Reply:
The cleanup is not needed for normal plugin since they are just "moved" from the tmp directory into their final destination. For meta plugin we explicitly create the meta directory in the plugins dir so we need to cleanup the temporary directory at the end (the plugins inside the meta directory are moved to their final destination but the root directory remains untouched).

Reply:
Hrm, I did wonder about that. Why is it we can't just move the meta plugin like a normal plugin?

Reply:
I refactored the code to move the meta plugin like a normal plugin:
https://github.com/elastic/elasticsearch/pull/28022/commits/e3083bad7f4e94951b6623bcbee064120a3c0ef2